src/CompoundRegion.cc

```diff
@@ -0,0 +1,196 @@
+    }
+    return result;
+}
+
+std::unique_ptr<CompoundRegion> CompoundRegion::decode(uint8_t const *buffer, size_t n) {
+    if (n == 0) {
+        throw std::runtime_error("Encoded CompoundRegion is truncated.");
+    }
+    switch (buffer[0]) {
+        case UnionRegion::TYPE_CODE:
+            return UnionRegion::decode(buffer, n);
+        case IntersectionRegion::TYPE_CODE:
+            return IntersectionRegion::decode(buffer, n);
+        default:
+            throw std::runtime_error("Byte string is not an encoded CompoundRegion.");
+    }
+}
+
+Box UnionRegion::getBoundingBox() const {
+    return getUnionBounds(*this, [](Region const &r) { return r.getBoundingBox(); });
+}
+
+Box3d UnionRegion::getBoundingBox3d() const {
+    return getUnionBounds(*this, [](Region const &r) { return r.getBoundingBox3d(); });
+}
+
+Circle UnionRegion::getBoundingCircle() const {
+    return getUnionBounds(*this, [](Region const &r) { return r.getBoundingCircle(); });
+}
+
+bool UnionRegion::contains(UnitVector3d const &v) const {
+    return getOperand(0).contains(v) || getOperand(1).contains(v);
+}
+
+Relationship UnionRegion::relate(Region const &rhs) const {
+    auto r1 = getOperand(0).relate(rhs);
+    auto r2 = getOperand(1).relate(rhs);
+    return
+        // Both operands must be disjoint with the given region for the union
+        // to be disjoint with it.
+        ((r1 & DISJOINT) & (r2 & DISJOINT))
+        // Both operands must be within the given region for the union to be
+        // within it.
+        | ((r1 & WITHIN) & (r2 & WITHIN))
+        // If either operand contains the given region, the union contains it.
+        | ((r1 & CONTAINS) | (r2 & CONTAINS));
+}
+
+Box IntersectionRegion::getBoundingBox() const {
+    return getIntersectionBounds(*this, [](Region const &r) { return r.getBoundingBox(); });
+}
+
+Box3d IntersectionRegion::getBoundingBox3d() const {
+    return getIntersectionBounds(*this, [](Region const &r) { return r.getBoundingBox3d(); });
+}
+
+Circle IntersectionRegion::getBoundingCircle() const {
+    return getIntersectionBounds(*this, [](Region const &r) { return r.getBoundingCircle(); });
+}
+
+bool IntersectionRegion::contains(UnitVector3d const &v) const {
+    return getOperand(0).contains(v) && getOperand(1).contains(v);
+}
+
+Relationship IntersectionRegion::relate(Region const &rhs) const {
+    auto r1 = getOperand(0).relate(rhs);
+    auto r2 = getOperand(1).relate(rhs);
+    return
+        // Both operands must contain the given region for the intersection to
+        // contain it.
+        ((r1 & CONTAINS) & (r2 & CONTAINS))
+        // If either operand is disjoint with the given region, the
+        // intersection is disjoint with it.
+        | ((r1 & DISJOINT) | (r2 & DISJOINT))
+        // If either operand is within the given region, the intersection is
+        // within it.
+        | ((r1 & WITHIN) | (r2 & WITHIN));
```

Comment:
I think there are also cases where neither operand is individually within rhs, but the intersection of the operands is (consider two overlapping circles, and rhs an ellipse that surrounds the intersection but extends outside both of the circles)

Reply:
(other thread on related problem for unions applies here, too)